### PR TITLE
[ISSUE #2001] support external cros filter config. 

### PR DESCRIPTION
**Pull request by qicz**
using `shenyu.cros-filter-config.*` config the cors filter:
```yml
shenyu:
  cros-filter-config:
    allowed-headers: qicz, custom
    allowed-expose: *
    allowed-methods: *
    allowed-origin: *
    max-age: 18000
```

the `allowed-headers` default is `x-requested-with, authorization, Content-Type, Authorization, credential, X-XSRF-TOKEN, token, username, client`.

 the external config will append to the `allowed-headers`.  the example of `allowed-headers` = `x-requested-with, authorization, Content-Type, Authorization, credential, X-XSRF-TOKEN, token, username, client,  qicz, custom`.

### CODE REVIEW

**Comment by yu199195** [comment]
@qicz HI , thanks for you pull request..  

cros config maybe in Shenyuconfig.java

**Comment by qicz** [comment]
> @qicz HI , thanks for you pull request..
> 
> cros config maybe in Shenyuconfig.java you can see this : #1999

hi,  this feature is not `hand code`. support external custom settings.

**Comment by yu199195** [comment]
> hi, this feature is not hand code. support external yml custom settings.

yes, yaml ----> Shenyuconfig.java , you can update code.....

**Comment by qicz** [comment]
> > hi, this feature is not hand code. support external yml custom settings.
> 
> yes, yaml ----> ShenyuConfig.java , you can update code.....

ok. I am not paying attention to it.

```java
@Bean
    @ConfigurationProperties(prefix = "shenyu")
    public ShenyuConfig shenyuConfig() {
        return new ShenyuConfig();
    }
```

Coding

**Comment by yu199195** [comment]
@qicz 

Hi , i think you can update code for master,  and then provide pull request to fix this function.

File: shenyu-web/src/main/java/org/apache/shenyu/web/configuration/properties/CrosFilterProperties.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * CrosFilterProperties
+ *
+ */
+@Component
+@ConfigurationProperties(prefix = "shenyu.cros")
+public class CrosFilterProperties {
```
Comment (yu199195) [inline shenyu-web/src/main/java/org/apache/shenyu/web/configuration/properties/CrosFilterProperties.java:18]: CrosFilterProperties maybe in ShenyuConfig.java ,not this  @ConfigurationProperties

File: shenyu-web/src/main/java/org/apache/shenyu/web/configuration/properties/ExcludePathProperties.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Rule out the url Filter.
+ */
+@Component
+@ConfigurationProperties(prefix = "shenyu.exclude")
```
Comment (yu199195) [inline shenyu-web/src/main/java/org/apache/shenyu/web/configuration/properties/ExcludePathProperties.java:30]: remove this...

File: shenyu-web/src/main/java/org/apache/shenyu/web/filter/TimeWebFilter.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.server.reactive.ServerHttpResponse;
+import org.springframework.web.server.ServerWebExchange;
+import org.springframework.web.server.WebFilterChain;
+import reactor.core.publisher.Mono;
+
+/**
+ * this is visit time verify filter.
+ */
+public class TimeWebFilter extends AbstractWebFilter {
```
Comment (yu199195) [inline shenyu-web/src/main/java/org/apache/shenyu/web/filter/TimeWebFilter.java:33]: this class removed in master

File: shenyu-web/src/main/java/org/apache/shenyu/web/configuration/properties/ExcludePathProperties.java (hunk 2 above)
Comment (yu199195) [reply shenyu-web/src/main/java/org/apache/shenyu/web/configuration/properties/ExcludePathProperties.java:30]: this class removed in master

File: shenyu-web/src/main/java/org/apache/shenyu/web/configuration/properties/CrosFilterProperties.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
+package org.apache.shenyu.web.configuration.properties;
```
Comment (yu199195) [inline shenyu-web/src/main/java/org/apache/shenyu/web/configuration/properties/CrosFilterProperties.java:1]: add apache license

File: shenyu-web/src/main/java/org/apache/shenyu/web/configuration/properties/CrosFilterProperties.java (hunk 1 above)
Comment (qicz) [reply shenyu-web/src/main/java/org/apache/shenyu/web/configuration/properties/CrosFilterProperties.java]: yes, I pull the new code from master.

File: shenyu-common/src/main/java/org/apache/shenyu/common/config/ShenyuConfig.java [hunk 5]
```diff
@@ -582,4 +608,159 @@ public void setPrintInterval(final Integer printInterval) {
...
+        public String getAllowedHeaders() {
+            return allowedHeaders = wrapperHeaders(allowedHeaders);
+        }
+
+        /**
+         * Sets the allowedHeaders.
+         *
+         * @param allowedHeaders allowedHeaders
+         */
+        public void setAllowedHeaders(String allowedHeaders) {
```
Comment (yu199195) [inline shenyu-common/src/main/java/org/apache/shenyu/common/config/ShenyuConfig.java:672]: add finnal

File: shenyu-web/src/main/java/org/apache/shenyu/web/configuration/properties/ExcludePathProperties.java (hunk 2 above)
Comment (qicz) [reply shenyu-web/src/main/java/org/apache/shenyu/web/configuration/properties/ExcludePathProperties.java:30]: > this class removed in master

copy

**Review by yu199195 — COMMENTED** [review]
add finnal to our method parameters

File: shenyu-web/src/main/java/org/apache/shenyu/web/controller/PluginController.java [hunk 6]
```diff
@@ -32,12 +32,7 @@
 import org.apache.shenyu.sync.data.api.PluginDataSubscriber;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.springframework.web.bind.annotation.GetMapping;
-import org.springframework.web.bind.annotation.PostMapping;
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RequestParam;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
```
Comment (yu199195) [inline shenyu-web/src/main/java/org/apache/shenyu/web/controller/PluginController.java:35]: can not modify this

**Comment by qicz** [comment]
fixed the cross filter enable switch.

```java
    @Bean
    @Order(-100)
    @ConditionalOnProperty(name = "shenyu.switchConfig.cross", havingValue = "true")
    public WebFilter crossFilter(final ShenyuConfig shenyuConfig) {
        return new CrossFilter(shenyuConfig.getCrosFilterConfig());
    }
```

**Review by yu199195 — APPROVED** [review]

**Pull request merged**